### PR TITLE
[#5054] improvement(api,server): Add the check of privileges

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?

Add the check of privileges

### Why are the changes needed?

Fix: #5054

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add some UTs

### CODE REVIEW

File: api/src/main/java/org/apache/gravitino/authorization/Privileges.java [hunk 1]
```diff
@@ -283,6 +297,16 @@ public static UseSchema allow() {
     public static UseSchema deny() {
       return DENY_INSTANCE;
     }
+
+    @Override
+    public boolean supportsMetadataObjectType(MetadataObject.Type type) {
+      Set<MetadataObject.Type> supportedTypes =
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/authorization/Privileges.java:303]: Make this a static variable.

File: api/src/main/java/org/apache/gravitino/authorization/Privileges.java [hunk 2]
```diff
@@ -327,10 +357,31 @@ public static CreateTable allow() {
     public static CreateTable deny() {
       return DENY_INSTANCE;
     }
+
+    @Override
+    public boolean supportsMetadataObjectType(MetadataObject.Type type) {
+      Set<MetadataObject.Type> supportedTypes =
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/authorization/Privileges.java:363]: Also here.

**Review by yuqi1129 — APPROVED** [review]
LGTM

**Comment by jerryshao** [comment]
I still have some comments, please wait for a while to merge.

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java [hunk 3]
```diff
@@ -195,18 +192,107 @@ private static void callAuthorizationPluginImpl(
...
+
+    Supplier<NoSuchMetadataObjectException> exceptionToThrowSupplier =
+        () ->
+            new NoSuchMetadataObjectException(
+                "Securable object %s doesn't exist", object.fullName());
+
+    switch (object.type()) {
+      case METALAKE:
+        check(
+            GravitinoEnv.getInstance().metalakeDispatcher().metalakeExists(identifier),
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java:218]: It is not good to use global static env variable here directly, which makes you hard to do the UT.

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java [hunk 4]
```diff
@@ -195,18 +192,107 @@ private static void callAuthorizationPluginImpl(
...
+            exceptionToThrowSupplier);
+        break;
+
+      default:
+        throw new IllegalArgumentException(
+            String.format("Doesn't support the type %s", object.type()));
+    }
+  }
+
+  public static void checkPrivilege(
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java:258]: I think we can combine this method and above together, so we don't have to access the DB twice.

File: api/src/main/java/org/apache/gravitino/authorization/Privilege.java [hunk 5]
```diff
@@ -39,6 +41,32 @@ public interface Privilege {
...
+   * @return It will return true if a metadata object supports the privilege otherwise false.
+   */
+  boolean supportsMetadataObjectType(MetadataObject.Type type);
+
+  /**
+   * @return the {@link SupportsSpecificCatalog} if the privilege only supports specific catalog
+   *     type.
+   * @throws UnsupportedOperationException if the privilege doesn't bind to a specific catalog type.
+   */
+  default SupportsSpecificCatalog supportsSpecificCatalog() {
```
[the remark this answers is not in the document]
Comment (jerryshao) [reply api/src/main/java/org/apache/gravitino/authorization/Privilege.java:57]: Or:

```java
default Optional<Catalog.Type> belongsToCatalog() {
  return Optional.empty()
}

File: api/src/main/java/org/apache/gravitino/authorization/Privilege.java [hunk 6]
```diff
@@ -39,6 +41,32 @@ public interface Privilege {
    */
   Condition condition();
 
+  /**
+   * If a metadata object supports the privilege, then this method will return true otherwise false.
+   *
+   * @param type The metadata object type.
+   * @return It will return true if a metadata object supports the privilege otherwise false.
+   */
+  boolean supportsMetadataObjectType(MetadataObject.Type type);
```
[the remark this answers is not in the document]
Comment (jerryshao) [reply api/src/main/java/org/apache/gravitino/authorization/Privilege.java:50]: Or `canApplyTo`

[the remark this answers is not in the document]
Comment (jerqi) [reply api/src/main/java/org/apache/gravitino/authorization/Privilege.java:50]: OK.

File: api/src/main/java/org/apache/gravitino/authorization/Privilege.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jerqi) [reply api/src/main/java/org/apache/gravitino/authorization/Privilege.java:57]: Prefer optional 1.

File: server/src/main/java/org/apache/gravitino/server/web/rest/RoleOperations.java [hunk 7]
```diff
@@ -118,12 +120,29 @@ public Response getRole(@PathParam("metalake") String metalake, @PathParam("role
...
       return Utils.doAs(
           httpRequest,
           () -> {
+            Set<MetadataObject> metadataObjects = Sets.newHashSet();
             for (SecurableObjectDTO object : request.getSecurableObjects()) {
-              checkSecurableObject(metalake, object);
+              MetadataObject metadataObject =
+                  MetadataObjects.parse(object.getFullName(), object.type());
+              if (metadataObjects.contains(metadataObject)) {
+                throw new IllegalArgumentException(
```
Comment (jerryshao) [inline server/src/main/java/org/apache/gravitino/server/web/rest/RoleOperations.java:132]: Why do we choose to throw an exception here, what is your design choice?

File: server/src/main/java/org/apache/gravitino/server/web/rest/RoleOperations.java [hunk 8]
```diff
@@ -118,12 +120,29 @@ public Response getRole(@PathParam("metalake") String metalake, @PathParam("role
...
+                    String.format(
+                        "Don't support duplicated metadata objects %s type %s",
```
Comment (jerryshao) [inline server/src/main/java/org/apache/gravitino/server/web/rest/RoleOperations.java:134]: Some places you're using securable objects, some places you're using metadata objects, I think you should unify them all.

**Comment by jerryshao** [comment]
I did some slight code polish, @jerqi @yuqi1129 please help to review.

**Review by jerryshao — APPROVED** [review]

**Review by yuqi1129 — APPROVED** [review]

**Pull request merged**